[김소라] IO와 NIO의 차이점은 무엇인가?

itsjustsora opened the pull request.

itsjustsora committed:
Create IO와 NIO.md

itsjustsora committed:
Merge remote-tracking branch 'origin/main'

youngreal commented on java/4-5주차/IO와 NIO.md:
```diff
@@ -0,0 +1,110 @@
+
+**2. Buffer**
+
+기존 IO에서도 Buffer로 한꺼번에 입력받고 출력하기 위해 BufferdInputStream과 같은 확장 클래스를 사용하기도 했는데
+NIO에서는 기본적으로 버퍼를 사용해서 입출력을 하기 때문에 효율적으로 처리가 가능하고 데이터의 위치를 이동하면서 필요한 부분을 읽고 쓸 수 있다.
```
버퍼를 사용하면 모아서 보내게 되어 효율적이라고 하셨는데
모아서 보내는게(버퍼를 사용하는게) 왜 효율적일까요?

youngreal commented on java/4-5주차/IO와 NIO.md:
```diff
@@ -0,0 +1,110 @@
+| 구분 | IO | NIO |
+| --- | --- | --- |
+| 입출력 방식 | 스트림 | 채널 |
+| 버퍼 방식 | 넌버퍼(Non-Buffer) | 버퍼(Buffer) |
```
다이렉트 버퍼 vs 넌 다이렉트 버퍼의 차이점도 추가해주시면 좋을것 같습니다

youngreal commented on java/4-5주차/IO와 NIO.md:
```diff
@@ -0,0 +1,110 @@
+<br>
+
+- 파일 다운로드 시 IO/NIO 중 선택할 방식과 이유
+> IO를 선택하겠습니다. 파일 다운로드 시에는 데이터를 읽어야만 저장할 수 있으므로 비동기적으로 처리해야 할 작업이 필요하지 않기 때문에 Non-Blocking으로 인한 추가 작업을 수행하지 않고 리소스를 절약하며 기능을 수행할 수 있을 것 같습니다.
```
버퍼의 크기는 얼마나 될 수 있을까요? 대용량 데이터를 버퍼를 이용하게되면 어떻게 될지도 생각해보시면 좋을것 같습니다

youngreal reviewed (commented):
NIO에서 기존의 문제를 해결하기위해 시스템콜 횟수 줄이기, 운영체제의 IO서비스(Scatter, Gather), 가상메모리등을 이용하게 되는데 위의 키워드가 어떻게 사용되는지  다시한번 살펴보시면 좋을것 같습니다.  
IO의 단점을 어떻게 개선한건지? 의 내용이겠네요

itsjustsora committed:
Merge branch 'Stacked-Book:main' into main

itsjustsora committed:
Update IO와 NIO.md
- Direct Buffer vs Non-Direct Buffer
- Scatter/Gather
- 가상 메모리
- File lock
- 버퍼 관련 답변 중 CPU 부하 내용 추가
